server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/ScanDataSource.java

```diff
@@ -90,25 +88,24 @@ class ScanDataSource implements DataSource {
 
   @Override
   public DataSource getNewDataSource() {
+
     if (isCurrent()) {
       return this;
     } else {
-      // log.debug("Switching data sources during a scan");
-      if (memIters != null) {
-        tablet.returnMemIterators(memIters);
-        memIters = null;
-        tablet.returnFilesForScan(fileReservationId);
-        fileReservationId = -1;
-      }
-
-      if (fileManager != null) {
-        tablet.getScanMetrics().decrementOpenFiles(fileManager.getNumOpenFiles());
-        fileManager.releaseOpenFiles(false);
+      try {
+        returnIterators();
+      } finally {
+        try {
+          if (fileManager != null) {
+            tablet.getScanMetrics().decrementOpenFiles(fileManager.getNumOpenFiles());
+            fileManager.releaseOpenFiles(false);
```

Comment:
releaseOpenFiles calls releaseReaders which can throw an IllegalArgumentException 

https://github.com/ivakegg/accumulo/blob/14dd12c3a97f44cd6c0273552ce8bc3b5439cd20/server/tserver/src/main/java/org/apache/accumulo/tserver/FileManager.java#L360

releaseOpenFiles is a synchronized method so it could get moved into returnIterators as releasing and decrementing the openFiles are actions performed by both close() and getNewDataSource()

Reply:
one of them calls releaseOpenFiles(false) and the other releaseOpenFiles(sawErrors).  Also note that this is within a try/finally block.

Reply:
The only difference is that `close()` sets the filemanager to null once it's done 
```
          if (fileManager != null) {
            tablet.getScanMetrics().decrementOpenFiles(fileManager.getNumOpenFiles());
            fileManager.releaseOpenFiles(false);
```

Reply:
Also one of them clears fileManager and the other does not.